### ISSUE TITLE
Kubernetes API Server should not depend on --cloud-provider and --cloud-config

### ISSUE THREAD

**Issue by dims** [kubernetes/kubernetes#49402]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
I was quite surprised to see `--cloud-config` and `--cloud-provider` in the api server when reviewing some code around openstack cloud provider. Turned out that this was needed by a single admission controller, namely `PersistentVolumeLabel`

**What you expected to happen**:

I expect kube-controller-manager and kubelet to depend on cloud-provider but NOT the api server.

**How to reproduce it (as minimally and precisely as possible)**:

```
[dims@bigbox 14:33] ~ ⟩ hyperkube apiserver --help 2>&1 | grep "cloud"
      --cloud-config string                                     The path to the cloud provider configuration file. Empty string for no configuration file.
      --cloud-provider string                                   The provider for cloud services. Empty string for no provider.
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by dims** [comment]
I poked at it a bit in this PR - https://github.com/kubernetes/kubernetes/pull/49374

**Comment by dims** [comment]
Code in question - https://github.com/kubernetes/kubernetes/blob/822e29dd3ce90421c099f6bec7f476c419c3211e/plugin/pkg/admission/persistentvolume/label/admission.go#L66

**Comment by dims** [comment]
one more dependency, when --external-hostname is not set in api server, we end up using the cloud provider to look for the external ip

```
(dlv) bt
 0  0x000000000253bdeb in k8s.io/kubernetes/pkg/cloudprovider/providers/openstack.getMetadataFromConfigDrive
    at ./_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/metadata.go:78
 1  0x000000000253cefa in k8s.io/kubernetes/pkg/cloudprovider/providers/openstack.getMetadata
    at ./_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/metadata.go:148
 2  0x0000000002541ba2 in k8s.io/kubernetes/pkg/cloudprovider/providers/openstack.(*Instances).CurrentNodeName
    at ./_output/local/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/openstack/openstack_instances.go:85
 3  0x00000000018cf001 in k8s.io/kubernetes/pkg/kubeapiserver/options.(*CloudProviderOptions).DefaultExternalHost
    at ./_output/local/go/src/k8s.io/kubernetes/pkg/kubeapiserver/options/cloudprovider.go:70
 4  0x0000000002ec4bb6 in k8s.io/kubernetes/cmd/kube-apiserver/app.defaultOptions
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:611
 5  0x0000000002ebf4df in k8s.io/kubernetes/cmd/kube-apiserver/app.CreateKubeAPIServerConfig
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:253
 6  0x0000000002ebe479 in k8s.io/kubernetes/cmd/kube-apiserver/app.CreateServerChain
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:127
 7  0x0000000002ebe34a in k8s.io/kubernetes/cmd/kube-apiserver/app.Run
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:113
 8  0x000000000380a876 in main.NewKubeAPIServer.func1
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/hyperkube/kube-apiserver.go:36
 9  0x0000000003807bdc in main.(*HyperKube).Run
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/hyperkube/hyperkube.go:177
10  0x0000000003808179 in main.(*HyperKube).RunToExit
    at ./_output/local/go/src/k8s.io/kubernetes/cmd/hyperkube/hyperkube.go:187
(dlv)
```

**Comment by smarterclayton** [comment]
Someone is in the process of removing this from sig-storage.  Not sure who.

**Comment by bgrant0607** [comment]
@dims 

I haven't had time to track it closely, but the Plan of Record is to get cloud providers out of tree entirely.

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/cloud-provider-refactoring.md

It should be possible to remove this particular dependency using initializers.

cc @kubernetes/sig-storage-bugs

**Comment by smarterclayton** [comment]
I believe that work is already in progress - @rrati didn't you have something open for this?

**Comment by dims** [comment]
Thanks @bgrant0607 @smarterclayton

**Comment by dims** [comment]
@smarterclayton @rrati Found it! looks like this one https://github.com/kubernetes/kubernetes/pull/44680

**Comment by luxas** [comment]
@dims Cool that you found it! I was just about to dig it up :)

**Comment by caesarxuchao** [comment]
cc @cheftako

**Comment by dims** [comment]
More work is being done in #51629

**Comment by luxas** [comment]
Also we should talk about if the SSH-to-nodes GCE thingy is really needed anymore with the latest auth features...
@kubernetes/sig-auth-feature-requests

**Comment by ericchiang** [comment]
@luxas is that involved with the cloud provider? I was under the impression it was just flags on the API server that didn't interact with an external service.

```
      --ssh-keyfile string                                      If non-empty, use secure SSH proxy to the nodes, using this user keyfile
      --ssh-user string                                         If non-empty, use secure SSH proxy to the nodes, using this user name
```

**Comment by luxas** [comment]
@ericchiang Now remember, I was thinking about the node dialer talking to the cloud provider re: installing SSH keys: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L212-L241

Only provider that implements this `AddSSHKeyToAllInstances` function in the cloudprovider interface is GCE. I'm asking whether this is _actually_ required anymore or if we can deprecate and remove...

**Comment by databus23** [comment]
@luxas Do you mean removing the ssh tunnelling feature completly or just removing `AddSSHKeyToAllInstances` from the cloud interface?

**Comment by liggitt** [comment]
I'd like to see the tunneling feature removed. If a particular deployment topology requires it, I would expect that deployment to set up VPN–style tunneling for specific IP ranges, rather than adding complexity throughout the API server

**Comment by databus23** [comment]
@liggitt Sorry to hear that. I was hoping to build on that feature in out deployment. Setting um vpn tunnels requires more outside machinery and complicates the setup considerably in my case.

**Comment by liggitt** [comment]
The complication is there regardless of whether it is built into the api server or not, and the tunneling built into the api server is opaque, not as robust as dedicated vpn solutions, and has been repeatedly problematic.

**Comment by jagosan** [comment]
@liggitt that seems like a reasonable direction to explore, but needs more discussion, as this is currently supported functionality. Also concerned that this comment will get buried and missed in what was reported as a bug.

**Comment by luxas** [comment]
@jagosan Anyway, it's only GCE that's using this. Removing GCE-only code in the core k8s codebase is a general theme anyway in my understanding. As @liggitt said, we don't strictly need this _in the core api server_.
On this note, kube-up.sh has been deprecated for a long time, and the plan is to move over to using kubeadm eventually. We will have to hack on the way GCE/GKE deploys clusters anyway. And, I'm not sure I see the exact value to have this now that we have better authn/authz mechanisms in core (as compared to before, where this was strictly needed)

Can we deprecate this functionality now, and work out a plan to stop using it in GCE/GKE gradually please?
We can definitely move this to a separate issue...

**Comment by jagosan** [comment]
@luxas I am not against the idea, just making the point that this discussion has veered off topic and don't want important discussions to get lost when this issue (reported as a bug with a title that isn't obviously related to the dialer's continued existence) is somehow resolved.

**Comment by dims** [comment]
This is part of larger work - https://github.com/kubernetes/kubernetes/issues/48690

**Comment by dims** [comment]
How to make progress from @lavalamp - https://github.com/kubernetes/kubernetes/pull/53912#issuecomment-337660614

**Comment by andrewsykim** [comment]
Yup we're working through this. I'm not sure we're at a place yet where we can remove that dependency given some providers still depend on it (e.g. `AddSSHKeyToAllInstances`). I think we're in agreeance that the SSH key feature should be removed (or at least shift responsibility to something that is not "core") but we need to be careful not to break anything. Our main focus right now is https://github.com/kubernetes/kubernetes/issues/69585, once that's done we should have some action items to tackle this.

**Comment by nikopen** [comment]
Greetings!
:mountain_snow: code freeze :mountain_snow:  is coming in about 10 days, is this intented to be implemented in the following 2-3 weeks?

@dims @andrewsykim @jagosan @cheftako

**Comment by andrewsykim** [comment]
@nikopen thanks for following up, this won't be in for v1.14.

**Comment by cheftako** [comment]
Removing this flag is reliant on completion of https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190226-network-proxy.md

**Comment by josiahbjorgaard** [comment]
Since this issue is tagged for milestone v1.16, here is a gentle reminder that code freeze is starting in 9 days on August 29th. Is the issue still targeted for this release?

**Comment by deads2k** [comment]
Since this issue was created, the in-tree cloud provider has been frozen and external cloud controller managers are active.  Once the transition is complete we'll be able to remove this.  I think the external cloud provider transition will continue with or without this issue and at this point there isn't a clear benefit to  transitioning the kube-apiserver early

/close
